File: 프로그래머스/lv3/최원준/코딩 테스트 공부.py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :
    dp 문제
2. 시간복잡도 :
    O( max_alp * max_cop * problems)
3. 자료구조 :
    배열
'''


def solution(alp, cop, problems):
    max_alp = max([problem[0] for problem in problems])
    max_cop = max([problem[1] for problem in problems])

    dp = [[float('inf')] * (max_cop + 1) for _ in range(max_alp + 1)]  # dp[alp][cop] = time
    alp, cop = min(alp, max_alp), min(cop, max_cop)
    dp[alp][cop] = 0

    for alp_score in range(alp, max_alp + 1):
        for cop_score in range(cop, max_cop + 1):
            if alp_score + 1 <= max_alp:  # +1 alp
                dp[alp_score + 1][cop_score] = min(dp[alp_score + 1][cop_score], dp[alp_score][cop_score] + 1)
            if cop_score + 1 <= max_cop:  # +1 cop
                dp[alp_score][cop_score + 1] = min(dp[alp_score][cop_score + 1], dp[alp_score][cop_score] + 1)

            for problem in problems:
                nalp, ncop, palp, pcop, time = problem

                if alp_score >= nalp and cop_score >= ncop:
                    ni = min(max_alp, alp_score + palp)
                    nj = min(max_cop, cop_score + pcop)
                    dp[ni][nj] = min(dp[ni][nj], dp[alp_score][cop_score] + time)

    return dp[max_alp][max_cop]

File: 프로그래머스/lv3/이재표/코딩테스트 공부.java
// 모든 문제를 풀수있는 알고력과 코딩력을 만드는데 드는 최단시간

import java.util.*;
class Solution {
    public int solution(int alp, int cop, int[][] problems) {
        int max_alp_req = 0;
        int max_cop_req = 0;
        
        for (int[] problem : problems) {
            max_alp_req = Math.max(max_alp_req, problem[0]);
            max_cop_req = Math.max(max_cop_req, problem[1]);
        }
        
        if (max_alp_req <= alp && max_cop_req <= cop) {
            return 0;
        }
        
        if (alp > max_alp_req) {
            alp = max_alp_req;
        }
        if (cop > max_cop_req) {
            cop = max_cop_req;
        }
        
        int[][] dp = new int[max_alp_req + 1][max_cop_req + 1];
        for (int i = 0; i < dp.length; i++) {
            for (int j = 0; j < dp[0].length; j++) {
                dp[i][j] = Integer.MAX_VALUE;
            }
        }
        
        dp[alp][cop] = 0;
        
        for (int i = alp; i <= max_alp_req; i++) {
            for (int j = cop; j <= max_cop_req; j++) {
                if (i < max_alp_req) {
                    dp[i + 1][j] = Math.min(dp[i + 1][j], dp[i][j] + 1);
                }
                if (j < max_cop_req) {
                    dp[i][j + 1] = Math.min(dp[i][j + 1], dp[i][j] + 1);
                }
                for (int[] problem : problems) {
                    if (i >= problem[0] && j >= problem[1]) {
                        int new_alp = Math.min(i + problem[2], max_alp_req);
                        int new_cop = Math.min(j + problem[3], max_cop_req);
                        dp[new_alp][new_cop] = Math.min(dp[new_alp][new_cop], dp[i][j] + problem[4]);
                    }
                }
            }
        }
        
        return dp[max_alp_req][max_cop_req];
    }
}
